Add documentation for different logpipeline comparison (#661)
kyma-project/community: concepts/observability-configurable-logging/07-multiple-logpipeline-investigation.md

# Objective

The telemetry operator comes with a pipeline configured to push logs to loki. A user of Kyma can define his own pipeline and push the logs to his own logging backend.

> The task here is to understand in a scenario when eg. the logging backend defined by  user is not working anymore what side effects it can have.

Reviewer: I unresolved this because it's still not clear whether it's an example or the only scenario.

### Case 2
![a](./assets/logpipeline-invstigation/case-2/case-2.svg)

Setup:
1. one input and 2 outputs (loki (fluentbit-loki plugin) + mockserver) with rewrite-tags
2. logpipelines: [loki](./assets/logpipeline-invstigation/case-2/loki.yaml), [mock-server](./assets/logpipeline-invstigation/case-2/mockserver.yml)

Result:
1. Output chunking keeps with 150M of newest data.
2.  The tail plugin sends the data to next phase of pipeline (filter plugin). If the outputs is not working then only the buffer at the rewrite tag would be filled and tail plugin buffer wont be filled
3. When the buffer is full then we see some error logs stating that and the old logs are discarded
    ```unix
    [2022/04/21 14:38:23] [error] [input:emitter:log_emitter] error registering chunk with tag: log_rewritten
    [2022/04/21 14:38:24] [error] [input chunk] chunk 1-1650551903.999375896.flb would exceed total limit size in plugin http.1
    ```
4. When grafana-loki plugin is stopped it stopped all the pipelines (so mockserver which was fine was also stopped) 
5. When mock server was down the logs were still shipped to loki

### Case 3

Setup:

> 1. one input with 2 outputs (mockserver + mockserver) with rewrite tags

Reviewer: ```suggestion
1. One input with two outputs (mock server and mock server) with rewrite tags
```

> 2. logpipelines: [mockserver-1](./assets/logpipeline-invstigation/case-3/mockserver-1.yml), [mockserver-2](./assets/logpipeline-invstigation/case-3/mockserver-2.yml)

Reviewer: ```suggestion
2. Two log pipelines:
   - [mockserver-1](./assets/logpipeline-invstigation/case-3/mockserver-1.yml)
   - [mockserver-2](./assets/logpipeline-invstigation/case-3/mockserver-2.yml)
```

> Result

Reviewer: ```suggestion
Result:
```

> 1. When both outputs are down then the buffer was filled and eventually the fluentbit pod got killed because of 500 (most probably because of CPU throttling)

Reviewer: If the order doesn't matter, use an unordered list

```suggestion
- When both outputs are down, the buffer was filled and eventually the Fluent Bit Pod went down with error code `500` (most probably because of CPU throttling).
```

> 2. Tail plugin kept pushing logs to rewrite buffer and they were eventually lost

Reviewer: ```suggestion
- Tail plugin kept pushing logs to the rewrite buffer, and they were eventually lost.
```

### Case 4
Setup:

> 1. one input with 2 outputs (loki (with official loki plugin) + mockserver) with rewrite tags

Reviewer: ```suggestion
1. One input with two outputs (Loki with the official Loki plugin, and mock server) with rewrite tags
```

> 2. logpipelines: [loki](./assets/logpipeline-invstigation/case-4/loki.yml), [mockserver](./assets/logpipeline-invstigation/case-4/mock-server.yml)

Reviewer: ```suggestion
2. Two log pipelines:
   - [Loki](./assets/logpipeline-invstigation/case-4/loki.yml)
   - [Mock server](./assets/logpipeline-invstigation/case-4/mock-server.yml)
```

> Result

Reviewer: ```suggestion
Result:
```

> 1. Mockserver was down and the loki output was still working

Reviewer: ```suggestion
- Mock server was down. The Loki output kept working.
```

> 2. The chunks in the filesystem buffer are rolled  (the old chunks deleted new ones created)

Reviewer: ```suggestion
- The chunks in the `filesystem` buffer are rolled: Old chunks are discarded; new ones created.
```

> 3. When loki output was down the mockserver output was still working

Reviewer: ```suggestion
- When Loki output was down, the mock server output kept working.
```

![mockserver-down](/assets/logpipeline-invstigation/case-4/dashboard-mock-down.png)

![loki-down](/assets/logpipeline-invstigation/case-4/dashboard-loki-down.png)

### Case 5
Setup:

> 1. one input with 2 outputs (loki (with official loki plugin) + mockserver) without rewrite tags

Reviewer: ```suggestion
1. One input with two outputs (Loki with the official Loki plugin, and mock server) without rewrite tags
```

> 2. logpipelines: [loki](./assets/logpipeline-invstigation/case-5/loki.yml), [mockserver](./assets/logpipeline-invstigation/case-5/mock-server.yml)

Reviewer: ```suggestion
2. Two log pipelines:
   - [Loki](./assets/logpipeline-invstigation/case-5/loki.yml)
   - [Mock server](./assets/logpipeline-invstigation/case-5/mock-server.yml)
```

> Result

Reviewer: ```suggestion
Result:
```

> 1. Mockserver was down, this led to filling of the tail buffer

Reviewer: If the order doesn't matter, use an unordered list

```suggestion
- Mock server was down. This led to the filling of the tail buffer.
```

> 2. When the tail plugin buffer is filled, it kept still reading.

Reviewer: ```suggestion
- After the tail plugin buffer was full, it kept reading.
```

> 3. The chunks in tail plugin is rolled (the old chunks deleted new ones created)

Reviewer: ```suggestion
- The chunks in the tail plugin buffer are rolled: Old chunks are discarded; new ones created.
```

> 4. Loki output stopped working as well

Reviewer: ```suggestion
- Loki output stopped working as well.
```

> 5. Found following [issue](https://github.com/fluent/fluent-bit/issues/4373) in github which describes the same problem.

Reviewer: That's not a result of the test though. It's a result of your research.
I suggest separating it from the results list with a blank line.

```suggestion

A known [Fluent Bit GitHub issue](https://github.com/fluent/fluent-bit/issues/4373) describes the same problem.
```

![mockserver-down](/assets/logpipeline-invstigation/case-5/dashboard-mock-down.png)

### Case 6

> Setup

Reviewer: ```suggestion
Setup:
```

> 1. 2 Inputs and 2 outputs (loki (fluentbit-loki plugin) + mockserver) and no rewrite tags

Reviewer: ```suggestion
- Two Inputs and two outputs (Loki (Fluent Bit-Loki plugin) and mock server) without rewrite tags
```

> 

Reviewer: Not mentioning the log pipelines for this setup?

Author: fixed.

Result

> 1. Each pipeline has its own buffer

Reviewer: ```suggestion
- Each pipeline has its own buffer.
```

> 2. When both the outputs are stopped both the tail plugins are loosing logs. It would keep the latest logs (although the amount of logs are differemt)

Reviewer: ```suggestion
- When both outputs are stopped, both tail plugins are losing logs. They keep the latest logs, but the amount of logs is different.
```
Just guessing; the sentences are a bit unclear to me. Who is "it"?

# Objective

The telemetry operator comes with a pipeline configured to push logs to Loki. Additionally, Kyma users can define their own pipeline and push the logs to their own logging backend.

> This document investigates the side effects that may come up when, for example, the logging backend defined by users stops working.

Author: This is more like setting up the context. Operator would have a pipeline preconfigured and user might want to add a new pipeline and we are investigating which approach of fluentbit config would be best suited for this.

    - Loki, which comes with Kyma
    - [Http output](./assets/logpipeline-investigation/mock-server.yaml)
3. [Log generator daemon set](./assets/logpipeline-investigation/log-generator.yaml) to generate a huge amount of logs to fill the filesystem buffer faster
4. A [Function](./assets/logpipeline-investigation/func.js) to check if the logs are being delivered when one of the outputs is having an outage
5. To simulate outage, the port of the service of the output was changed so that the DNS resolution keeps working but logs won't be delivered.

Behaviour of fluent-bit is different when using `in-memory` buffer and `filesystem` buffer. With `in-memory` buffer the input is paused if an output is having an outage. Whereas in case of `filesystem` buffer the logs are still read by the tail plugin and stored in the filesystem buffer until the output is available again. However, logs are dropped if the fileystem buffer is full.

## Test Cases

### Setup 1
This was not tested because of high resource overhead and increased complexity of the implementation of the telemetry operator.

### Setup 2

![a](./assets/logpipeline-investigation/setup-2/setup-2.drawio.svg)

Setup:

> - Two Inputs and two outputs (Loki (grafana-loki plugin) and http output (http plugin)) without rewrite tags

Reviewer: ```suggestion
- Two inputs and two outputs (Loki (grafana-loki plugin) and http output (http plugin)) without rewrite tags
```

> - Two logpipelines: 

Reviewer: ```suggestion
- Two log pipelines: 
```

>   - [loki](./assets/logpipeline-investigation/setup-2/loki.yml)

Reviewer: ```suggestion
  - [Loki](./assets/logpipeline-investigation/setup-2/loki.yml)
```

  - [Http output](./assets/logpipeline-investigation/setup-2/mock-server.yml)
- [telemetry config](./assets/logpipeline-investigation/setup-2/telemetry-config)

Result
- Each pipeline has its own fileystem buffer.
- When both outputs are having outage, both tail plugins are losing logs. The buffer in the tail plugin stores the latest logs, but the amount of logs is different.
### Setup 3a
![a](./assets/logpipeline-investigation/setup-3a/setup-3a.drawio.svg)

Setup:

> 1. One input with two outputs (Loki (grafana-loki plugin) and Http output (http plugin)) with rewrite tags (with filesystem buffer)

Author: The plugin name is `grafana-loki` thats how we are using it in the config. Would refrain from renaming this as it might confuse?

> 1. One input with two outputs (Loki (grafana-loki plugin) and Http output (http plugin)) with rewrite tags (with filesystem buffer)

Reviewer: oh, good point. then we might need the backticks. I thought we're referring to the proper product names here

2. Two log pipelines:

>    - [Loki](./assets/logpipeline-investigation/setup-3a/loki.yaml)

Reviewer: ```suggestion
   - [Loki](./assets/logpipeline-investigation/setup-3a/loki.yml)
```

A known [Fluent Bit GitHub issue](https://github.com/fluent/fluent-bit/issues/4373) describes the same problem.

### Setup 4
![a](./assets/logpipeline-investigation/setup-4/setup-4.drawio.svg)

Setup:
1. One input with two outputs (limiting the maximum number of chunks in the `filesystem` buffer) without rewrite tags
2. Two log pipelines:
   - [Loki](./assets/logpipeline-investigation/setup-4/loki.yaml)
   - [Http output](./assets/logpipeline-investigation/setup-4/mockserver.yml)

Result:
- When one of the outputs is having an outage, the filesystem buffer is filled up and only keeps 150M of latest data (old data is discarded first). Eventually, when the filesystem buffer at tail plugin is full, the output plugin for Loki stops as well.
- When both outputs are having an outage, the tail plugin is stopped after filling the filesystem buffer, even though the buffer was not completely full (104M/150M).

## Summary

> We performed various tests and found that with filesystem buffering (through rewrite_tag filter) is necessary to prevent loss of logs when fluent-bit pod is restarted. Additionally when one of output is having an outage, then the logs are still streamed to the other output. However loss of logs cannot be prevented completely, if the buffer is filled up then fluentbit keeps only the latest logs.

Reviewer: I'm not completely sure about the meaning of the first sentence. We need a comma, and maybe an "it". The meanings will be different so I don't know which one is correct.

```suggestion
We performed various tests and found that with filesystem, buffering (through rewrite_tag filter) is necessary to prevent loss of logs when fluent-bit pod is restarted. Additionally, when one of the outputs is having an outage, the logs are still streamed to the other output. However, loss of logs cannot be prevented completely: If the buffer is filled up, Fluent Bit keeps only the latest logs.
```

or 

```suggestion
We performed various tests and found that with filesystem buffering (through rewrite_tag filter), it is necessary to prevent loss of logs when fluent-bit pod is restarted. Additionally, when one of the outputs is having an outage, the logs are still streamed to the other output. However, loss of logs cannot be prevented completely: If the buffer is filled up, Fluent Bit keeps only the latest logs.
```

Additionally following case wes tried

> 1. With 2 input and 2 outputs and having `storage.max_chunks_pause on`. This option as mentioned in [documentation](https://docs.fluentbit.io/manual/administration/buffering-and-storage#input-section-configuration) did not apply to input filters. The fluentbit did not recongnize this option and the pod would not start. However applying to the service section did not have any affect.

Reviewer: ```suggestion
1. With two inputs and two outputs and having `storage.max_chunks_pause on`. This option, as mentioned in [documentation](https://docs.fluentbit.io/manual/administration/buffering-and-storage#input-section-configuration), did not apply to input filters. The Fluent Bit did not recognize this option and the Pod did not start. However, applying the setup to the service section did not have any affect.
```
Just guessing the last sentence's meaning

# Objective

The telemetry operator comes with a pipeline configured to push logs to Loki. Additionally, Kyma users can define their own pipeline and push the logs to their own logging backend.

This document investigates the side effects that may come up when, for example, the logging backend defined by  user stops working.

## Setup

> The investigated configuration uses the `filesystem buffering`. When one of the log pipeline outputs fails, the logs are buffered on the file system. In contrast, the `in-memory` buffer has the disadvantage that if the Fluent Bit Pod is restarted, logs can be lost. During the test, we referred to the Fluent Bit knowledge article about [Buffering & Storage](https://docs.fluentbit.io/manual/administration/buffering-and-storage).

Reviewer 2: Let's use "file system" or "filesystem" coherently

Reviewer 2: We should also mention that Fluent Bit behaves differently when the filesystem storage is active:
- If the non-persistent memory storage is used, the input is paused if an output stops processing.
- If the filesystem storage is used, the input is never stopped and the filesystem utilized if logs cannot be send to the output. Logs are dropped if the configured limit of the filesystem storage is reached.

Author: fixed

The setup consisted of following items:
1. Kyma with telemetry operator
2. Two outputs, deployed in the Kyma cluster:
    - Loki, which comes with Kyma

>     -[Mock server](./assets/logpipeline-invstigation/mock-server.yaml)

Reviewer 2: missing space

3. [Log generator daemon set](./assets/logpipeline-invstigation/log-generator.yaml) to generate a huge amount of logs to fill the filesystem buffer faster
4. A [Function](./assets/logpipeline-invstigation/func.js) to check if the logs are being delivered when one of the outputs is down
5. To simulate failures, the port of the service was changed so that the DNS resolution keeps working but logs won't be delivered.

## Test Cases

### Case 1
![a](./assets/logpipeline-invstigation/case-1/case-1.svg)

Setup:
1. One input with two outputs (limiting the maximum number of chunks in the `filesystem` buffer) without rewrite tags
2. Two log pipelines:
   - [Loki](./assets/logpipeline-invstigation/case-1/loki.yaml)
   - [Mock server](./assets/logpipeline-invstigation/case-1/mockserver.yml)

Result:

> - When one of the outputs is down, the filesystem buffer is filled up and only keeps 150M of latest data (old data is discarded first). Eventually, when the filesystem buffer at tail plugin is full, the output plugin for Loki stops as well.

Reviewer 2: Be more precise what "output is down" means. Make clear that this is not related to the Fluent Bit configuration, but and outage of an external system.

Author: now I am using outage to make it more explicit.

- When both outputs are down, the tail plugin is stopped after filling the filesystem buffer, even though the buffer was not completely full (104M/150M).

### Case 2
![a](./assets/logpipeline-invstigation/case-2/case-2.svg)

Setup:

> 1. One input with two outputs (Loki (grafana-loki plugin) and mock server (http plugin)) with rewrite tags

Reviewer 2: Worth to mention that both of the rewrite_tag filters have their own filesystem storage.

   - [mockserver-2](./assets/logpipeline-invstigation/case-3/mockserver-2.yml)

![a](./assets/logpipeline-invstigation/case-3/case-3.svg)
Result:
- When both outputs are down, the filesystem buffer was filled and eventually the Fluent Bit Pod went down with error code `500` (most probably because of CPU throttling).
- Tail plugin kept pushing logs to the rewrite filesystem buffer, and they were eventually lost.
### Case 4
Setup:
1. One input with two outputs (Loki (Loki plugin), and mock server (http plugin)) with rewrite tags
2. Two log pipelines:
   - [Loki](./assets/logpipeline-invstigation/case-4/loki.yml)
   - [Mock server](./assets/logpipeline-invstigation/case-4/mock-server.yml)

![a](./assets/logpipeline-invstigation/case-4/case-4.svg)

Result:
- Mock server was down. The Loki output kept working.
- The chunks in the `filesystem buffer` are rolled: Old chunks are discarded; new ones created.
- When Loki output was down, the mock server output kept working.

![mockserver-down](/assets/logpipeline-invstigation/case-4/dashboard-mock-down.png)

![loki-down](/assets/logpipeline-invstigation/case-4/dashboard-loki-down.png)

> ### Case 5

Reviewer 2: What is the difference to case 1? The other Loki Plugin? Since Case 2-5 are very similar (one input with rewrite_tag filter), maybe we could make it Case 2a, Case 2b and Case 2c. That would give a structure that is a bit easier to follow

Reviewer 2: Maybe we could also keep the naming of the previous document:
- Setup 1 not tested here
- Setup 2 = Case 6
- Setup 3a-c = Case 2-6
- Setup 4 =  Case 1

Author: renamed.

- After the tail plugin filesystem buffer was full, it kept reading.
- The chunks in the tail plugin filesystem buffer are rolled: Old chunks are discarded; new ones created.
- Loki output stopped working as well.
- The logs are stored in filesystem buffer before applying the kuberenets filter. There is a risk of not being able to fetch the kubernetes metadata of the logs. This applies if the pod has been terminated before the logs have been read from the filesystem again for next stage of pipeline processing.

A known [Fluent Bit GitHub issue](https://github.com/fluent/fluent-bit/issues/4373) describes the same problem.

### Case 6
Setup:
- Two Inputs and two outputs (Loki (grafana-loki plugin) and mock server (http plugin)) without rewrite tags
- Two logpipelines: 
  - [loki](./assets/logpipeline-invstigation/case-6/loki.yml)
  - [mockserver](./assets/logpipeline-invstigation/case-6/mock-server.yml)
![a](./assets/logpipeline-invstigation/case-6/case-6.svg)

Result
- Each pipeline has its own fileystem buffer.
- When both outputs are stopped, both tail plugins are losing logs. They keep the latest logs, but the amount of logs is different.

## Summary

> We performed various tests and found that with filesystem buffering (through rewrite_tag filter) is necessary to prevent loss of logs when one of the output is down. However loss of logs cannot be prevented completely, if the buffer is filled up then fluentbit keeps only the latest logs.

Reviewer 2: I the statement of the first sentence is not correct. Filesystem buffering helps us to prevent loosing logs when restarting Fluent Bit. The advantage is that a broken output would not stall all outputs.

Author: fixed.

Additionally following case wes tried
1. With 2 input and 2 outputs and having `storage.max_chunks_pause on`. This option as mentioned in [documentation](https://docs.fluentbit.io/manual/administration/buffering-and-storage#input-section-configuration) did not apply to input filters. The fluentbit did not recongnize this option and the pod would not start. However applying to the service section did not have any affect.

> ## Conclusion

Reviewer 2: Maybe we could move the conclusion back to the other document to have only the technical description of the test cases here. That would make it one combined concept and not just a collection of unrelated files

Author: moved.

A known [Fluent Bit GitHub issue](https://github.com/fluent/fluent-bit/issues/4373) describes the same problem.

### Setup 4
![a](./assets/logpipeline-investigation/setup-4/setup-4.drawio.svg)

Setup:
1. One input with two outputs (limiting the maximum number of chunks in the `filesystem` buffer) without rewrite tags
2. Two log pipelines:
   - [Loki](./assets/logpipeline-investigation/setup-4/loki.yaml)
   - [Http output](./assets/logpipeline-investigation/setup-4/mockserver.yml)

Result:
- When one of the outputs is having an outage, the filesystem buffer is filled up and only keeps 150M of latest data (old data is discarded first). Eventually, when the filesystem buffer at tail plugin is full, the output plugin for Loki stops as well.
- When both outputs are having an outage, the tail plugin is stopped after filling the filesystem buffer, even though the buffer was not completely full (104M/150M).

## Summary

> We performed various tests and found that with filesystem, buffering (through rewrite_tag filter) is necessary to prevent loss of logs when fluent-bit pod is restarted. Additionally, when one of the outputs is having an outage, the logs are still streamed to the other output. However, loss of logs cannot be prevented completely: If the buffer is filled up, Fluent Bit keeps only the latest logs.

Reviewer: I must have missed this earlier.

```suggestion
We performed various tests and found that with filesystem, buffering (through rewrite_tag filter) is necessary to prevent loss of logs when the Fluent Bit Pod is restarted. Additionally, when one of the outputs is having an outage, the logs are still streamed to the other output. However, loss of logs cannot be prevented completely: If the buffer is filled up, Fluent Bit keeps only the latest logs.
```

A known [Fluent Bit GitHub issue](https://github.com/fluent/fluent-bit/issues/4373) describes the same problem.

### Setup 4

![a](./assets/logpipeline-investigation/setup-4/setup-4.drawio.svg)

Setup:
1. One input with two outputs (limiting the maximum number of chunks in the `filesystem` buffer) without rewrite tags
2. Two log pipelines:
   - [Loki](./assets/logpipeline-investigation/setup-4/loki.yaml)
   - [Http output](./assets/logpipeline-investigation/setup-4/mockserver.yml)

Result:
- When one of the outputs is having an outage, the filesystem buffer is filled up and only keeps 150M of latest data (old data is discarded first). Eventually, when the filesystem buffer at tail plugin is full, the output plugin for Loki stops as well.
- When both outputs are having an outage, the tail plugin is stopped after filling the filesystem buffer, even though the buffer was not completely full (104M/150M).

## Summary

> We performed various tests and found that with filesystem, buffering (through rewrite_tag filter) is necessary to prevent loss of logs when the Fluent Bit Pod is restarted. Additionally, when one of the outputs is having an outage, the logs are still streamed to the other output. However, loss of logs cannot be prevented completely: If the buffer is filled up, Fluent Bit keeps only the latest logs.

Reviewer: Having read the whole doc in order, I think we need to put the comma to a different place and add the "it". 
Please correct me if I'm misunderstanding the meaning of this sentence.

```suggestion
We performed various tests and found that with filesystem buffering (through rewrite_tag filter), it is necessary to prevent loss of logs when the Fluent Bit Pod is restarted. Additionally, when one of the outputs is having an outage, the logs are still streamed to the other output. However, loss of logs cannot be prevented completely: If the buffer is filled up, Fluent Bit keeps only the latest logs.
```

Additionally following case wes tried
1. With 2 input and 2 outputs and having `storage.max_chunks_pause on`. This option as mentioned in [documentation](https://docs.fluentbit.io/manual/administration/buffering-and-storage#input-section-configuration) did not apply to input filters. The fluentbit did not recongnize this option and the pod would not start. However applying to the service section did not have any affect.

## Conclusion
After trying out various possibilities of log pipeline, we decided to choose go with Case 4 for following reasons:
1. In the event of failure of one output, fluent-bit would still push logs to other output.
2. The rewrite tag with filesystem buffer enables the logs to buffered as chunks in case of failure of one outputs.

however, in case of failure of output for prolonged period of time and the filesystem buffer being full, there is loss of logs seen.

A decision is still needed how the rewrite tag should be configured. There are following proposals:

> 1. The user configures it himself if needed. Documentation should be provided to use rewrite tag as recommendation for the use cases where logs need to be sent to multiple logging backends along with the example how to do it.

Author: fixed

> 2. The rewrite tag along with name would be configured dynamically when the user creates a new pipeline.

Reviewer: ```suggestion
- The rewrite tag, along with name, is configured dynamically when the user creates a new pipeline.
```
Is configured by whom? Not by the user themselves (as in the other option). We should be clearer on this.

Author: fixed.
